src/algebra/homology/homological_complex.lean

```diff
@@ -0,0 +1,646 @@
+arrow.hom_mk (f.comm_from i)
+
+@[simp] lemma sq_from_left (f : hom C₁ C₂) (i : ι) : (f.sq_from i).left = f.f i := rfl
+@[simp] lemma sq_from_right (f : hom C₁ C₂) (i : ι) : (f.sq_from i).right = f.f_next i := rfl
+
+@[simp] lemma sq_from_id (C₁ : homological_complex V c) (i : ι) : sq_from (𝟙 C₁) i = 𝟙 _ :=
+begin
+  rcases h : c.next i with _ | ⟨j,w⟩,
+  { ext,
+    { refl, },
+    { dsimp, simp only [f_next, h],
+      symmetry,
+      apply zero_of_target_iso_zero,
+      exact X_next_iso_zero _ h, }, },
+  { ext, refl, dsimp, simp [f_next, h], }
+end
+
+@[simp] lemma sq_from_comp (f : C₁ ⟶ C₂) (g : C₂ ⟶ C₃) (i : ι) :
+  sq_from (f ≫ g) i = sq_from f i ≫ sq_from g i :=
+begin
+  rcases h : c.next i with _ | ⟨j,w⟩,
+  { ext,
+    { refl, },
+    { dsimp, simp only [f_next, h],
+      symmetry,
+      apply zero_of_target_iso_zero,
+      exact X_next_iso_zero _ h, }, },
+  { ext, refl, dsimp, simp [f_next, h], }
+end
+
+/--
+A morphism of chain complexes
+induces a morphism of arrows of the differentials into each object.
+-/
+def sq_to (f : hom C₁ C₂) (j : ι) : arrow.mk (C₁.d_to j) ⟶ arrow.mk (C₂.d_to j) :=
+arrow.hom_mk (f.comm_to j)
+
+@[simp] lemma sq_to_right (f : hom C₁ C₂) (j : ι) : (f.sq_to j).right = f.f j := rfl
+
+end hom
+
+end homological_complex
+
+namespace chain_complex
+
+section of
+variables {V}
+
+/--
+Construct a `ℕ`-indexed chain complex from a dependently-typed differential.
+-/
+def of (X : ℕ → V) (d : Π n, X (n+1) ⟶ X n) (sq : ∀ n, d (n+1) ≫ d n = 0) : chain_complex V ℕ :=
+{ X := X,
+  d := λ i j, if h : i = j + 1 then
+    eq_to_hom (by subst h) ≫ d j
+  else
+    0,
+  shape' := λ i j w, by rw dif_neg (ne.symm w),
+  d_comp_d' := λ i j k,
+  begin
+    split_ifs with h h' h',
+    { substs h h',
+      simp only [category.id_comp, eq_to_hom_refl],
+      exact sq k, },
+    all_goals { simp },
+  end, }
+
+variables (X : ℕ → V) (d : Π n, X (n+1) ⟶ X n) (sq : ∀ n, d (n+1) ≫ d n = 0)
+
+@[simp] lemma of_X (n : ℕ) : (of X d sq).X n = X n := rfl
+@[simp] lemma of_d (j : ℕ) : (of X d sq).d (j+1) j = d j :=
+by { dsimp [of], rw [if_pos rfl, category.id_comp], }
+
+end of
+
+section mk
+
+/--
+Auxiliary structure for setting up the recursion in `mk`.
+This is purely an implementation detail: for some reason just using the dependent 6-tuple directly
+results in `mk_aux` taking about much longer (well over the `-T100000` limit) to elaborate.
+-/
+@[nolint has_inhabited_instance]
+structure mk_struct :=
+(X₀ X₁ X₂ : V)
+(d₀ : X₁ ⟶ X₀)
+(d₁ : X₂ ⟶ X₁)
+(s : d₁ ≫ d₀ = 0)
+
+variables {V}
+
+/-- Flatten to a tuple. -/
+def mk_struct.flat (t : mk_struct V) :
+  Σ' (X₀ X₁ X₂ : V) (d₀ : X₁ ⟶ X₀) (d₁ : X₂ ⟶ X₁), d₁ ≫ d₀ = 0 :=
+⟨t.X₀, t.X₁, t.X₂, t.d₀, t.d₁, t.s⟩
+
+variables (X₀ X₁ X₂ : V) (d₀ : X₁ ⟶ X₀) (d₁ : X₂ ⟶ X₁) (s : d₁ ≫ d₀ = 0)
+  (succ : Π (t : Σ' (X₀ X₁ X₂ : V) (d₀ : X₁ ⟶ X₀) (d₁ : X₂ ⟶ X₁), d₁ ≫ d₀ = 0),
+    Σ' (X₃ : V) (d₂ : X₃ ⟶ t.2.2.1), d₂ ≫ t.2.2.2.2.1 = 0)
```

Comment:
This `t.2.2.2.2.1` makes me think it might be better to just use `mk_struct` instead of the flat version. Is there any benefit to using a sigma type?

Reply:
@adamtopaz I don't have sufficient overview, but I feel like Scott need sigma types heavily for construction projective resolutions etc...